mail/protocols/imap/src/test/java/com/fsck/k9/mail/store/imap/ImapResponseParserTest.kt

```diff
@@ -0,0 +1,579 @@
+        }.isFailure()
+            .isInstanceOf(NegativeImapResponseException::class)
+    }
+
+    @Test
+    fun `readResponse() with resp-text code containing a list`() {
+        val parser = createParserWithResponses(
+            "* OK [PERMANENTFLAGS (\\Answered \\Flagged \\Deleted \\Seen \\Draft NonJunk \$MDNSent \\*)] " +
+                "Flags permitted.",
+        )
+
+        val response = parser.readResponse()
+
+        assertThat(response).hasSize(3)
+        assertThat(response).index(0).isEqualTo("OK")
+        assertThat(response).index(1).isInstanceOf(ImapList::class).all {
+            index(0).isEqualTo("PERMANENTFLAGS")
+            index(1).isInstanceOf(ImapList::class).containsExactly(
+                "\\Answered",
+                "\\Flagged",
+                "\\Deleted",
+                "\\Seen",
+                "\\Draft",
+                "NonJunk",
+                "\$MDNSent",
+                "\\*",
+            )
+        }
+
+        assertThat(response).index(2).isEqualTo("Flags permitted.")
+        assertThatAllInputWasConsumed()
+    }
+
+    @Test
+    fun `readResponse() with untagged EXISTS response`() {
+        val parser = createParserWithResponses("* 23 EXISTS")
+
+        val response = parser.readResponse()
+
+        assertThat(response).hasSize(2)
+        assertThat(response).transform { it.getNumber(0) }.isEqualTo(23)
+        assertThat(response).transform { it.getString(1) }.isEqualTo("EXISTS")
+        assertThatAllInputWasConsumed()
+    }
+
+    @Test(expected = IOException::class)
+    fun `readResponse() should throw if stream ends before end of line is found`() {
+        val parser = createParserWithData("* OK Some text ")
+
+        parser.readResponse()
+    }
+
+    @Test
+    fun `readResponse() with command continuation`() {
+        val parser = createParserWithResponses("+ Ready for additional command text")
+
+        val response = parser.readResponse()
+
+        assertThat(response.isContinuationRequested).isTrue()
+        assertThat(response).containsExactly("Ready for additional command text")
+        assertThatAllInputWasConsumed()
+    }
+
+    @Test
+    fun `readResponse() with literal`() {
+        val parser = createParserWithResponses("* {4}\r\ntest")
+
+        val response = parser.readResponse()
+
+        assertThat(response).containsExactly("test")
+        assertThatAllInputWasConsumed()
+    }
+
+    @Test
+    fun `readResponse() with empty literal`() {
+        val parser = createParserWithResponses("* {0}\r\n")
+
+        val response = parser.readResponse()
+
+        assertThat(response).containsExactly("")
+        assertThatAllInputWasConsumed()
+    }
+
+    @Test(expected = IOException::class)
+    fun `readResponse() should throw when end of stream is reached while reading literal`() {
+        val parser = createParserWithData("* {4}\r\nabc")
+
+        parser.readResponse()
+    }
```

Comment:
I think we should stick to one failure testing method and use assertk here too.

```suggestion
    @Test
    fun `readResponse() should throw when end of stream is reached while reading literal`() {
        val parser = createParserWithData("* {4}\r\nabc")

        assertThat {
            parser.readResponse()
        }.isFailure()
            .isInstanceOf(IOException::class)
    }
```

Reply:
That was my intention. Thanks for catching the instances I missed :+1:

Fixed in 22e499301db6ba54147a3c55169ab3d9ebe635d8